### ISSUE TITLE
uwsgi + pypy :  wsgi.input.read truncates http body by null char

### ISSUE THREAD

**Issue by cutewalker** [unbit/uwsgi#918]
with uwsgi on pypy,  envrion['wsgi.input'].read would truncates http body by null char.
for example, we post binary data with content type application/octet-stream, null char is not mean to the end. 

after I check the source, I find that the problem may be caused by misuse of ffi.string, as described in [the doc](https://cffi.readthedocs.org/en/latest/#misc-methods-on-ffi).

and i think maybe ffi.buffer is better: 

```
diff --git a/plugins/pypy/pypy_setup.py b/plugins/pypy/pypy_setup.py
index ac83c75..7e4be6e 100644
--- a/plugins/pypy/pypy_setup.py
+++ b/plugins/pypy/pypy_setup.py
@@ -393,7 +393,8 @@ class WSGIinput(object):
         rlen = ffi.new('ssize_t*')
         chunk = lib.uwsgi_request_body_read(self.wsgi_req, size, rlen)
         if chunk != ffi.NULL:
-            return ffi.string(chunk, rlen[0])
+            #return ffi.string(chunk, rlen[0])
+            return ffi.buffer(chunk, rlen[0])[:]
         if rlen[0] < 0:
             raise IOError("error reading wsgi.input")
         raise IOError("error waiting for wsgi.input")
```

**Comment by unbit** [comment]
Do you still get problems with this patch applied or it solves your issue ?

**Comment by cutewalker** [comment]
yeah,  the patch solves my issue

**Comment by unbit** [comment]
this misuse of ffi.string was everywhere. I have fixed it in 2.0.11. Thanks

**Comment by cutewalker** [comment]
oh, it's my pleasure.
